### ISSUE TITLE
Mypy, circular imports, and working directory

### ISSUE THREAD

**Issue by adrianparvino** [flycheck/flycheck#1829]
## Checklist
- [/] I have checked existing issues for potential duplicates before creating this one.
- [ ] I have read the [Troubleshooting guide][].

## Bug description
Mypy is meant to be run from the project root. Otherwise, circular imports will occur.

## Steps to reproduce
1. Create a standard Python package,
2. Import a package with the same name as the file, e.g.:
foo.py
```python
import foo

...
```
3. Source file found twice under different module names: 'project.foo' and 'foo'

- [ ] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior
This should typecheck properly.

## System configuration
```
Syntax checkers for buffer celery.py in python-mode:

First checker to run:

  python-flake8
    - may enable:         yes
    - executable:         Found at /nix/store/lix6bnvjr9wah1fjaswlsvb50gjsdp4y-python3-3.8.5-env/bin/python3
    - configuration file: Not found
    - `flake8' module:    Found at "/nix/store/lix6bnvjr9wah1fjaswlsvb50gjsdp4y-python3-3.8.5-env/lib/python3.8/site-packages/flake8/__init__.py"
    - next checkers:      python-pylint, python-mypy

Checkers that may run as part of the first checker's chain:

  python-mypy
    - may enable:         yes
    - may run:            t
    - executable:         Found at /nix/store/lix6bnvjr9wah1fjaswlsvb50gjsdp4y-python3-3.8.5-env/bin/mypy
    - configuration file: Found at "/home/myrl/Development/aesdata/remote_server/tasks/mypy.ini"

Checkers that could run if selected:

  python-pycompile  select
    - may enable:    yes
    - executable:    Found at /nix/store/lix6bnvjr9wah1fjaswlsvb50gjsdp4y-python3-3.8.5-env/bin/python3
    - next checkers: python-mypy

Checkers that are compatible with this mode, but will not run until properly configured:

  python-pylint (automatically disabled) reset
    - may enable:         no
    - executable:         Found at /nix/store/lix6bnvjr9wah1fjaswlsvb50gjsdp4y-python3-3.8.5-env/bin/python3
    - configuration file: Not found
    - `pylint' module:    Missing; sys.path is ['/nix/store/bs03sg8b0gq2zr4v252hh9psp780qj5q-python3-3.8.5/lib/python38.zip', '/nix/store/bs03sg8b0gq2zr4v252hh9psp780qj5q-python3-3.8.5/lib/python3.8', '/nix/store/bs03sg8b0gq2zr4v252hh9psp780qj5q-python3-3.8.5/lib/python3.8/lib-dynload', '/nix/store/bs03sg8b0gq2zr4v252hh9psp780qj5q-python3-3.8.5/lib/python3.8/site-packages', '/nix/store/lix6bnvjr9wah1fjaswlsvb50gjsdp4y-python3-3.8.5-env/lib/python3.8/site-packages']
    - next checkers:      python-mypy

  python-pyright (automatically disabled) reset
    - may enable: no
    - executable: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot
Emacs version:    28.0.50
System:           x86_64-pc-linux-gnu
Window system:    x
```

Emacs configuration:
- [/] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes
The closest related issue is #1711. The solution there does not fix this, but the external solution https://github.com/lbolla/emacs-flycheck-mypy/commit/d5c791f does.

Mypy is needed to be run from the project's root. I've applied the patch locally, and it works.

[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by smdnv** [comment]
I'm having the same issue. When mypy is run not from project root it returns next error:
```
```

**Comment by smdnv** [comment]
I'm having the same issue. When mypy is run not from project root it returns next error:
```
~project/project_package/project_module
❯ /home/user/.virtualenvs/project123/bin/mypy \
--show-column-numbers \
--config-file /home/user/projects/project123/setup.cfg \
/home/user/projects/project123/project_package/project_module/enum.py
enum.py: error: Source file found twice under different module names: 'project_package.project_module.enum' and 'enum'
Found 1 error in 1 file (checked 1 source file)
```

When I'm running mypy to check the same file but from project root it returns no error:
```
~/project
❯ /home/user/.virtualenvs/project123/bin/mypy \
--show-column-numbers \
--config-file /home/user/projects/project123/setup.cfg \
/home/user/projects/project123/project_package/project_module/enum.py/home/user/projects/project123/project_package/project_module/enum.py
Success: no issues found in 1 source file
```
This command I took from `flycheck-compile`.

Another problem that I noticed mypy creates .mypy_cache in each folder where a file was checked, for example:
```
~/project
❯ find . | grep -E "mypy.*flask$"
./project_package/project_module/module1/.mypy_cache/3.8/flask
./project_package/project_module/module1/module2/.mypy_cache/3.8/flask
./.mypy_cache/3.8/flask
```

When I have checked files in a project for few days, this cache used a lot of memory ~1Gb.
```
~/project
❯ du -h  | grep -E 'mypy_cache$' | sort -h
# ...
23M     ./project_package1/module1/.mypy_cache
24M     ./project_package2/module3/module4/.mypy_cache
# ...
```

P.s. Sorry for those clumsy names, I tested it on closed project.

**Comment by adrianparvino** [comment]
Hi, I created a patch to apply this.

CC: @smdnv 

```
From f7a0f0c2ec25b4a2d5ef548fe87d5c5c5d43c183 Mon Sep 17 00:00:00 2001
From: "Adrian Parvin D. Ouano" <me@adrianpar.vin>
Date: Wed, 27 Jan 2021 03:49:08 +0800
Subject: [PATCH] Run flycheck in root

---
 flycheck.el | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/flycheck.el b/flycheck.el
index 8d0b3e16..331c34f7 100644
--- a/flycheck.el
+++ b/flycheck.el
@@ -10671,6 +10671,14 @@ See URL https://github.com/microsoft/pyright."
   :safe #'flycheck-string-or-nil-p
   :package-version '(flycheck . "32"))
 
+
+(defun flycheck-mypy--find-project-root (_checker)
+  "Compute an appropriate working-directory for flycheck-mypy.
+This is either a parent directory containing a flycheck-mypy.ini, or nil."
+  (and
+   buffer-file-name
+   (locate-dominating-file buffer-file-name "mypy.ini")))
+
 (flycheck-define-checker python-mypy
   "Mypy syntax and type checker.  Requires mypy>=0.580.
 
@@ -10680,6 +10688,7 @@ See URL `http://mypy-lang.org/'."
             (config-file "--config-file" flycheck-python-mypy-config)
             (option "--cache-dir" flycheck-python-mypy-cache-dir)
             source-original)
+  :working-directory flycheck-mypy--find-project-root
   :error-patterns
   ((error line-start (file-name) ":" line (optional ":" column)
           ": error:" (message) line-end)
-- 
2.28.0
```

**Comment by smdnv** [comment]
@adrianparvino 
Thank you for your example, it's working great.

P.s. Advice for other users: If you are using doom, don't forget to run `doom build` to recompile the package and apply changes.

**Comment by wbolster** [comment]
i ran into the same issue, and opened a pull request for it: https://github.com/flycheck/flycheck/pull/1878

please leave a comment there to help the reviewers.

my solution is roughly the same as what @adrianparvino shared before (and similar to what various other checkers do)

**Comment by jap** [comment]
Ran into this as well, and can confirm that applying #1878 makes mypy work brilliantly!

**Comment by bbatsov** [comment]
This has been fixed in the current codebase. The `python-mypy` checker now uses `flycheck-python-find-project-root` as its `:working-directory`, which resolves to the project root containing `setup.cfg`, `pyproject.toml`, or similar project files. This should resolve the circular import issues caused by running mypy from the wrong directory.
